### ISSUE TITLE
[RISCV] Backend hangs during `RISCV DAG->DAG Pattern Instruction Selection` when having binary ops after `select`

### ISSUE THREAD

**Issue by HazyFish** [llvm/llvm-project#58511]
## Description

The backend hangs during `RISCV DAG->DAG Pattern Instruction Selection` pass for the following input IR when targeting `riscv32` / `riscv64`. 

The constant operands for `select` must be -1 and 0 (order doesn't matter) to reproduce the problem.
The problem also occurs when `mul` is replaced by `add` / `sub`, but not `udiv` / `sdiv` / `urem` / `srem`.
The problem also occurs when `or` is replaced by `and`, but not `xor`.
The constant used in the 2 binary ops must be greater than or equal to 2048 and the constant operand for `mul` must be equal or slightly greater than the constant operand used for `or` (difference needs to be below 2048) to reproduce the problem.

## Cause

[`DAGCombiner::foldBinOpIntoSelect`](https://github.com/llvm/llvm-project/blob/4309bb28ae77061d528b09dfe1c41335e92bc7a3/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L2228) and [`DAGCombiner::foldSelectOfBinops`](https://github.com/llvm/llvm-project/blob/4309bb28ae77061d528b09dfe1c41335e92bc7a3/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L24342) are called for infinite number of times and the DAG is alternating between 2 forms as shown below:

```
SelectionDAG has 19 nodes:
  t0: ch,glue = EntryToken
        t4: i32,ch = CopyFromReg t0, Register:i32 %1
      t13: i32 = mul t4, OpaqueConstant:i32<2048>
      t6: i32,ch = CopyFromReg t0, Register:i32 %2
    t16: ch = store<(store (s32) into %ir.2)> t0, t13, t6, undef:i32
          t2: i32,ch = CopyFromReg t0, Register:i32 %0
        t7: i1 = truncate t2
      t12: i32 = select t7, Constant:i32<-1>, Constant:i32<0>
    t14: i32 = or t12, OpaqueConstant:i32<2048>
  t18: ch,glue = CopyToReg t16, Register:i32 $x10, t14
  t19: ch = RISCVISD::RET_FLAG t18, Register:i32 $x10, t18:1

SelectionDAG has 20 nodes:
  t0: ch,glue = EntryToken
        t4: i32,ch = CopyFromReg t0, Register:i32 %1
      t13: i32 = mul t4, OpaqueConstant:i32<2048>
      t6: i32,ch = CopyFromReg t0, Register:i32 %2
    t16: ch = store<(store (s32) into %ir.2)> t0, t13, t6, undef:i32
        t2: i32,ch = CopyFromReg t0, Register:i32 %0
      t7: i1 = truncate t2
      t20: i32 = or Constant:i32<-1>, OpaqueConstant:i32<2048>
      t21: i32 = or Constant:i32<0>, OpaqueConstant:i32<2048>
    t22: i32 = select t7, t20, t21
  t18: ch,glue = CopyToReg t16, Register:i32 $x10, t22
  t19: ch = RISCVISD::RET_FLAG t18, Register:i32 $x10, t18:1

SelectionDAG has 19 nodes:
  t0: ch,glue = EntryToken
        t4: i32,ch = CopyFromReg t0, Register:i32 %1
      t13: i32 = mul t4, OpaqueConstant:i32<2048>
      t6: i32,ch = CopyFromReg t0, Register:i32 %2
    t16: ch = store<(store (s32) into %ir.2)> t0, t13, t6, undef:i32
          t2: i32,ch = CopyFromReg t0, Register:i32 %0
        t7: i1 = truncate t2
      t23: i32 = select t7, Constant:i32<-1>, Constant:i32<0>
    t24: i32 = or t23, OpaqueConstant:i32<2048>
  t18: ch,glue = CopyToReg t16, Register:i32 $x10, t24
  t19: ch = RISCVISD::RET_FLAG t18, Register:i32 $x10, t18:1

SelectionDAG has 20 nodes:
  t0: ch,glue = EntryToken
        t4: i32,ch = CopyFromReg t0, Register:i32 %1
      t13: i32 = mul t4, OpaqueConstant:i32<2048>
      t6: i32,ch = CopyFromReg t0, Register:i32 %2
    t16: ch = store<(store (s32) into %ir.2)> t0, t13, t6, undef:i32
        t2: i32,ch = CopyFromReg t0, Register:i32 %0
      t7: i1 = truncate t2
      t25: i32 = or Constant:i32<-1>, OpaqueConstant:i32<2048>
      t26: i32 = or Constant:i32<0>, OpaqueConstant:i32<2048>
    t27: i32 = select t7, t25, t26
  t18: ch,glue = CopyToReg t16, Register:i32 $x10, t27
  t19: ch = RISCVISD::RET_FLAG t18, Register:i32 $x10, t18:1

SelectionDAG has 19 nodes:
  t0: ch,glue = EntryToken
        t4: i32,ch = CopyFromReg t0, Register:i32 %1
      t13: i32 = mul t4, OpaqueConstant:i32<2048>
      t6: i32,ch = CopyFromReg t0, Register:i32 %2
    t16: ch = store<(store (s32) into %ir.2)> t0, t13, t6, undef:i32
          t2: i32,ch = CopyFromReg t0, Register:i32 %0
        t7: i1 = truncate t2
      t28: i32 = select t7, Constant:i32<-1>, Constant:i32<0>
    t29: i32 = or t28, OpaqueConstant:i32<2048>
  t18: ch,glue = CopyToReg t16, Register:i32 $x10, t29
  t19: ch = RISCVISD::RET_FLAG t18, Register:i32 $x10, t18:1

...
```

### Related Code

https://github.com/llvm/llvm-project/blob/4309bb28ae77061d528b09dfe1c41335e92bc7a3/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L24364-L24373

https://github.com/llvm/llvm-project/blob/4309bb28ae77061d528b09dfe1c41335e92bc7a3/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L2283-L2304

## Minimal Reproduction

https://godbolt.org/z/5xE3c6xjb

### Code

```ll
define i32 @f(i1 %0, i32 %1, ptr %2) {
BB:
  %I = select i1 %0, i32 -1, i32 0
  %I1 = mul i32 %1, 2048
  %I2 = or i32 2048, %I
  store i32 %I1, ptr %2
  ret i32 %I2
}
```

**Comment by HazyFish** [comment]
cc @DataCorrupted

**Comment by HazyFish** [comment]
@topperc I have updated the description to include more details. Hope it's helpful

**Comment by topperc** [comment]
Should be fixed after 00816714f96505c59c236f3f0fd2eb815c57f674
